keps/sig-api-machinery/1904-efficient-watch-resumption/README.md

```diff
@@ -0,0 +1,485 @@
+
+* **Will enabling / using this feature result in increasing time taken by any 
+operations covered by [existing SLIs/SLOs]?**
+  Think about adding additional work or introducing new steps in between
+  (e.g. need to do X to start a container), etc. Please describe the details.
+
+* **Will enabling / using this feature result in non-negligible increase of 
+resource usage (CPU, RAM, disk, IO, ...) in any components?**
+  Things to keep in mind include: additional in-memory state, additional
+  non-trivial computations, excessive access to disks (including increased log
+  volume), significant amount of data sent and/or received over network, etc.
+  This through this both in small and large cases, again with respect to the
+  [supported limits].
+
+### Troubleshooting
+
+The Troubleshooting section currently serves the `Playbook` role. We may consider
+splitting it into a dedicated `Playbook` document (potentially with some monitoring
+details). For now, we leave it here.
+
+_This section must be completed when targeting beta graduation to a release._
+
+* **How does this feature react if the API server and/or etcd is unavailable?**
+
+* **What are other known failure modes?**
+  For each of them, fill in the following information by copying the below template:
+  - [Failure mode brief description]
+    - Detection: How can it be detected via metrics? Stated another way:
+      how can an operator troubleshoot without logging into a master or worker node?
+    - Mitigations: What can be done to stop the bleeding, especially for already
+      running user workloads?
+    - Diagnostics: What are the useful log messages and their required logging
+      levels that could help debug the issue?
+      Not required until feature graduated to beta.
+    - Testing: Are there any tests for failure mode? If not, describe why.
+
+* **What steps should be taken if SLOs are not being met to determine the problem?**
+
+[supported limits]: https://git.k8s.io/community//sig-scalability/configs-and-limits/thresholds.md
+[existing SLIs/SLOs]: https://git.k8s.io/community/sig-scalability/slos/slos.md#kubernetes-slisslos
+
+## Implementation History
+
+2020-06-30: KEP Proposed.
+
+## Drawbacks
+
+n/a
+
+## Alternatives
+
+### Initialize watch cache from etcd history window
+
+The main alternative to the above solution would be to change the way how
+watch cache is initialized to not only initialize the state but also a
+transaction log (i.e. read the whole etcd history and initialize transaction
+log based on it).
+
+Pros:
+- doesn't require any etcd changes
+- no overhead when kube-apiserver is running (only initialization being more
+   expensive)
+
+Cons:
+- given kube-apiserver is performing compaction (default every 5m), lack of
+   changes of any object of type X within that period would result in inability
+   to initialize transaction log anyway; so that is not universal solution
```

Comment:
Isn't getting the latest RV the most valuable part, though?

Reply:
That doesn't solve the second problem though (i.e. you can have multiple kube-apiservers out-of-sync for extended period of time).

Based on your comments above, I think you misunderstood that second problem - so hopefully with new clarifications it's clearer what I mean here.

Reply:
Ah, so the actual need is to select the oldest RV that is greater than any change to the objects in the collection.

If it weren't for deleted objects not showing up in the list, it would be trivial, you wouldn't even need to read the change history.

But if we can read the change history, I don't agree with this con. If you can get:
* the list of changes from the history window since the last compaction, and it is empty
* C = RV of the last compaction
* O = greatest RV of an object returned by list
* G = global RV as of the time of the list
and if O < C <= G, then it is safe to pretend that you've synced to C instead of G. (given that the list is empty, this will always be the case. If the list is not empty you can still use C as the RV as long as you fill the revision history in memory with the list.)

This seems like a much less time-sensitive fix than relying on not being about to generate the condition in 250ms.

It may be reasonable to make both fixes, but if we pre-fill the cache like this, then we just need to give the client an RV newer than the last compaction, which is drastically less frequent than 250ms.

Reply:
> Ah, so the actual need is to select the oldest RV that is greater than any change to the objects in the collection.

Probably equal to the most fresh change instead of "great than any change to the objects in this collection" (we need to be able to say from which point in time we can start serving watches and be sure that nothing has happened after that). 
But yeah - this is +-1 difference, so probably doesn't matter.


> the list of changes from the history window since the last compaction, and it is empty

This one is tricky. Because even if we would have a way to expose RV of the last compaction, the all changes since then we can only get via watch. And if we establish watch, we have no way to say if it already synced to now or not...
We can probably solve that with using ProgressNotify again, but I think we would end-up with similarly if not more complicated solution (with the solution I'm proposing there aren't any corner cases)...

Reply:
OK, here's a very stupid idea that is probably easy to implement and effective.

When filling the watch cache, currently we start watching for changes at the index given by the list.

Imagine a change that, in addition to doing that, also "binary" searches backwards in time with a series of watch requests.

E.g.:
* watch zero is at RV=X. Cache is now synced to X. We let this watch continue. (same watch we do today.)
* watch 1 is at RV=X-1000. We get an out-of-window error.
* watch 2 is at RV=X-500. We get 3 changes with RVs < X; we prepend them to our change list and now our cache is synced to X-500. Close the watch.
* watch 3 is at RV = X-750, out of window.
* watch 4 is at RV = X-625, we get 1 additional change, prepend it, cache is now synced at RV=X-625.

4 iterations is probably more than enough, but you can use this technique to find the exact limit (which might be a moving target!).

Reply:
Yeah - I've been thinking about that too. But imagine there are no changes after RV=X.
So when you initialized watch2 from RV=X-500 - how you can answer whether it already reached X or not?

Reply:
Well, not getting an out-of-window error tells us something?

We could enable progress-notify on these extra watches :)

Reply:
> We could enable progress-notify on these extra watches :)

Sure - but with that we need to do most of changes described in the doc anyway...

I guess what I'm saying is that - sure we can do that, but it's actually not that much simpler (if at all).
[Note that we would need to be able to present history, which we don't support now, so another complication to what people, including Jordan already claim to be terribly complicated...]

Reply:
It is a change, and it makes the watch system more complicated, but it's localized and doesn't change any client-visible interfaces. In that sense it is well contained. We can make sure the testing is good. Personally I feel like making apiserver load prior history adds a lot of robustness to the system, and making the system depend on something happening every 250ms adds brittleness.

But it may be time for @deads2k and @liggitt to weigh in.

Reply:
If we would really require these 250ms for something - true. But we can tolerate much more (given kube-apiserver initializes tens of seconds anyway).

Reply:
Or I should say it a bit differently. In order to utilize the progress notify to detect whether watch already synced to RV=X, we basically need to bump the frequency to something reasonable from the current 10m anyway. So pretty much all the changes described in this proposal would have to be implemented anyway.

Reply:
Or let me actually write it differently:
I'm not opposed to changing how we initialize watchcache, but based on all the arguments above (mostly need for watch progress notify to implement it, the fact that we can still have two apiservers claimed to be synced to different RVs for extended period of time), I would treat it as a complementary not a replacement of what I'm proposing.

Reply:
The frequency change requires etcd work? Or is it just a setting when you start a watch?

I'm uncomfortable with the solution being strongly time dependent. I currently still think I'd like apiserver to pre-load enough items into the cache to make the chance of a race extremely low. If we can make the "race window" 10s instead of 250ms, I think I'm fine with that.

Reply:
> The frequency change requires etcd work? Or is it just a setting when you start a watch?

Yes - currently progress notify is hardcoded to be sent every 10m.

>  If we can make the "race window" 10s instead of 250ms, I think I'm fine with that.

I already wrote above that 250ms doesn't seem to be needed. I was thinking about something between 1s and 5s. But maybe 10s is good enough too? What we need is to ensure that in rolling-upgrade scenario, the next-to-be-updated kube-apiserver will get a progress-notify event *after* the previous is initialized.

So 10s sounds fine to me to - given we want to send bookmarks on shutdown we need to be able to catch the signal and propagate it down. So we can also wait for the next progress notify in that scenario.

Actually - the above are important details about how this can/should be tuned. I added it to the proposed solution above (in the Proposal section).

Reply:
@lavalamp - hopefully this addresses your concern?:
https://github.com/kubernetes/enhancements/pull/1878/files#diff-d73a72730139197a054ffcd22466d771R243-R254

Reply:
OK, yeah, that should keep it from being racy.